### ISSUE TITLE
Can't select a value on iOS and Android

### ISSUE THREAD

**Issue by Peppe** [vaadin/vaadin-combo-box#646]
### Description
Values can't be easily selected. This happens in all combo boxes. I noticed it first in Bakery while I tried to fill in a form, and reproduced it in component browser. This happens when you use either iOS or Android. Workaround is to click on the dropdown icon, which doesn't open the keyboard. With that the selection work.

### Steps to reproduce
1. Whip out your phone
1. Go to https://vaadin.com/components/vaadin-combo-box/html-examples
1. Touch on the combo-box's text "Carbon". It opens up the dropdown, while on the same time opens up the virtual keyboard of your phone.
1. Try to select "Boron" from the list

### Expected outcome
The value "Boron" becomes selected.

### Actual outcome
Clicking on "Boron" closes the keyboard, closes the popup, and keeps "Carbon selected.

### Browsers Affected
- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] IE 11
- [X] iOS Safari
- [X] Android Chrome

**Comment by jouni** [comment]
I also tried to reproduce this, and the problem only appears if the combo box that you focus is below a certain threshold in the viewport so that the viewport has to be scrolled up to center the field within the remaining viewport height. If the field is already in the correct place, everything works fine.

Also, the fact that the combo box is not actually modal inside a vaadin-dialog (the same issue as https://github.com/vaadin/vaadin-dropdown-menu/issues/120) causes the tap to hit the wrong target and focus an unexpected field in certain cases.

**Comment by jouni** [comment]
Here’s what happens:

### No viewport scroll (combo boxes are high enough in the viewport):

![combo-box-no-viewport-scroll](https://user-images.githubusercontent.com/66382/39305583-4bc7d528-4966-11e8-9ff9-d18dd3dc7ec0.gif)

### Viewport scrolls (combo boxes are lower inside the viewport):

![combo-box-viewport-scroll](https://user-images.githubusercontent.com/66382/39305606-5fbb29b8-4966-11e8-97e3-d34b65df8896.gif)

**Comment by jouni** [comment]
Can’t reproduce on CDN.

**Comment by web-padawan** [comment]
> It opens up the dropdown, while on the same time opens up the virtual keyboard of your phone.

As a user, I would expect to see the keyboard opening when clicking on something which looks as a text field. This is also needed to filter the items in the combo-box.

We need to investigate the viewport scroll, though.

**Comment by Peppe** [comment]
My use case here.
![combobox](https://user-images.githubusercontent.com/525503/39310422-9eaa21ac-496a-11e8-9552-fcc5444c0fdc.gif)

**Comment by jouni** [comment]
> > It opens up the dropdown, while on the same time opens up the virtual keyboard of your phone.

> As a user, I would expect to see the keyboard opening when clicking on something which looks as a text field. This is also needed to filter the items in the combo-box.

Yes, that’s the designed and expected behavior, and I’m not considering to change that. As Jens mentioned, you can open the dropdown, without opening the keyboard, by pressing the toggle button.

**Comment by tomivirkki** [comment]
I don't see this as an issue. I'm ok with closing.

**Comment by jouni** [comment]
Point 4 seems incorrect. In the recordings we shared, the overlay is still under the field. But clicks on the items in the overlay somehow end up targeting incorrect elements which are visually above the overlay.

**Comment by web-padawan** [comment]
Now I get it, thanks. Will continue investigating.

**Comment by jouni** [comment]
Still reproducible, although doesn’t seem that easy anymore.

I used Simulator, iPhone 8, iOS 11.2. Go to the Users view, edit some user, try to change the role of the user. Sometimes the combo box doesn’t choose the value that you tapped on.

**Comment by manolo** [comment]
### Summary

1. @Peppe  issue seems not being reproducible any more like is showed in this video

http://recordit.co/F85OptFnlc

2. There is different issue that happens very randomly as described by @jouni when opening and closing several times the combo-box, but anyways failures are very rare.

Take a look to this video http://recordit.co/eJoFtlaSfM in second 0:33 there is a failure, and in 0:48 a couple of them

Cause seems to be that when clicking on an item a different element is targeted. Perhaps related to the fact that the overlay is moved using transform-translate. We might try a different approach, but its not easy to reproduce and test 

3. In addition I have observed another issue that happens when we click on an element that is in the area covered by the navigation bar when opened, we need double touching/click the element as it showed in the following videos in order to fire its action.

http://recordit.co/Li7jsxqLyx

It's not a problem in the combo-box but a feature in iOS that happens with any other clickable component as demonstrated in this video

http://recordit.co/N6XsnORQrG

### Actions

  - Should we close this issue ?
  - Or do we try to fix 2.?

**Comment by manolo** [comment]
Pushed a PR for discussion about UX change proposal. It fixes problems described in this issue since the keyboard is not opened unless the user needs to filter content, and he needs to touch a second time on the input field (when the overlay is opened). @jouni wdyt? 

![double-touch-combo-box](https://user-images.githubusercontent.com/161853/40159558-27da5998-59aa-11e8-895f-0e05845a526c.gif)

**Comment by jouni** [comment]
My first impression is positive. Have to test it, though.

I’m not 100% sure about the versioning implications. I would like to say this is not a breaking change and can be introduced in a minor version, but it might also cause a big change in some user flows in existing apps.

**Comment by jouni** [comment]
About your point `#3` (nav bar area): that’s a known iOS limitation/behavior and not related to this issue, like you said. I would not think it affects the issue described here.

I’m wondering what kind of test setup you have for this? My assumption is that it’s a combination of having the combo box inside a fixed positioned container, possibly even a scrollable fixed positioned container.

Regarding the proposed new behavior, my fear is that there is no indication for the end user that they can actually type in the field – they would have to know that or accidentally discover it. Which sounds bad 😕

**Comment by jouni** [comment]
I forgot to add, that one thing I believe could be causing problems is that we close the keyboard immediately when the user interacts with the overlay. It might make sense to actually close the overlay only when scrolling (touchmove vs touchstart). That’s my assumption, didn’t actually check in the code.

**Comment by manolo** [comment]
Overlay is closed when
   - blurring the input
   - touch-end on the overlay because it's supposed the user selected an item

Soft keyboard is closed when 
   - blurring the input
   - touch-end on the overlay
   - touch-move on the overlay because it's supposed the user is going to scroll

**Comment by manolo** [comment]
> we might consider to disable search feature if the number of items is under N

@jouni wdyt on that idea

For example its BFF it is really annoying that the keyboard is opened in order just to be able to filter items, but there are only 3 options.
It's actually faster to click on the correct option than try to write a filter.  Note that allowCustomValue is disabled.

So probably a good UX might be that when certain conditions match (touchdevice, no customvalue, few items) we prevent using the keyboard

**Comment by jouni** [comment]
I agree that the keyboard is annoying for short lists. I think it makes sense to have some threshold of items.

The conditions you propose sound good. I would set the item limit to 10 (those still fit in without scrolling on iPhone 6/7/8).

**Comment by manolo** [comment]
Another option is to add a new property like `no-filter`, so as when `no-filter` and not `allow-custom-value` happen.
In this case there are no breaking changes and it's developer who decides whether activate default filtering.
